### ISSUE TITLE
Consider linting :paths value data type in bb.edn and deps.edn

### ISSUE THREAD

**Issue by lread** [clj-kondo/clj-kondo#1353]
### Problem
On more than one occasion I have included the following in my `bb.edn` file:
```Clojure
{:paths [script]
```
Although this is syntactically incorrect, babashka happens to parse it just fine. 
Other tools might not be so lenient or even crash. For example, it was [causing clojure-lsp to crash](https://github.com/clojure-lsp/clojure-lsp/issues/507).

The correct syntax is a vector of strings:
```Clojure
{:paths ["script"]
```

It would be helpful if clj-kondo warned when the data type is incorrect for `:paths`.

### Additional context
There is an existing [`:deps.edn` linter](https://github.com/clj-kondo/clj-kondo/blob/master/doc/linters.md#depsedn) which already does some data type checking, ex. for 
```Clojure
{:deps {foo/bar "2020.10.11"}}
``` 
clj-kondo reports:
```shell
deps.edn:1:18: warning: Expected map, found: java.lang.String
```
(Aside: this particular check does not seem to currently apply to `bb.edn`)

### Proposed Solution

Follow existing `:deps.edn` linter (see above) convention and show "Expected &lt;data type x&gt;, found: &lt;data type y&gt;." In this specific case:
`expected java.lang.String, found symbol`
or maybe:
`expected string, found symbol`

### Questions
Should we bundle`bb.edn` checking into the existing `:deps.edn` linter? Or should we create a new `:bb.edn` linter?

Should we expand scope of this issue?:
- in addition to checking for string, also check that `:paths` specifies a vector?
- lint data types of other elements in `bb.edn`?
- also likewise lint `:paths` in `deps.edn`? 
  - I do see a [spec](https://github.com/clojure/tools.deps.alpha/blob/9bf5778dc26dd5018dbf04fc8e7dbb32ddc4036c/src/main/clojure/clojure/tools/deps/alpha/specs.clj) for it
  - Note that in `deps.edn`, `:paths` can be strings or keywords (aliases).

###  Action
There are open questions, if it makes sense, we can move a GitHub discussion and I can come back and edit this issue with the outcome.
After we agree on what we want to do, I can follow up with  a PR.

**Comment by borkdude** [comment]
> Should we bundlebb.edn checking into the existing :deps.edn linter? Or should we create a new :bb.edn linter?

I think we can re-use the same code, but disabling some checking when we are in bb.edn.

> in addition to checking for string, also check that :paths specifies a vector?

Yes, or at least a seqable of strings (or keywords if those are allowed, but symbols seem to be out of the question).

**Comment by borkdude** [comment]
Additional specific bb.edn checks should go into a different issue I think.

**Comment by lread** [comment]
Thanks, sounds good. I'll take a crack at the PR sometime soon.

**Comment by lread** [comment]
So I've done some poking around, @borkdude. You are most certainly aware of all this, but I needed to learn/explore.

### deps.edn :paths with aliases
It seems like Babashka and Clojure will both pick up alias keywords in `:paths` from `deps.edn`.

In a new empty directory, given the following `deps.edn`
```Clojure
{:paths ["src" "test" :aliasa :aliasb]
 :aliases {:aliasa ["a1" "a2"]
           :aliasb ["a3"]}}
```

From Clojure, we get the following classpath (notice `src:test:a1:a2:a3`)
```Shell
❯ clojure -Spath
src:test:a1:a2:a3:/Users/lee/.m2/repository/org/clojure/clojure/1.10.3/clojure-1.10.3.jar:/Users/lee/.m2/repository/org/clojure/core.specs.alpha/0.2.56/core.specs.alpha-0.2.56.jar:/Users/lee/.m2/repository/org/clojure/spec.alpha/0.2.194/spec.alpha-0.2.194.jar
```

From Babashka, we get the following classpath (notice `src:test:a1:a2:a3`)
```Shell
❯ bb print-deps --format classpath
src:test:a1:a2:a3:/Users/lee/.m2/repository/babashka/babashka.curl/0.0.3/babashka.curl-0.0.3.jar:/Users/lee/.m2/repository/babashka/fs/0.0.5/fs-0.0.5.jar:/Users/lee/.m2/repository/cheshire/cheshire/5.10.1/cheshire-5.10.1.jar:/Users/lee/.m2/repository/clj-commons/clj-yaml/0.7.107/clj-yaml-0.7.107.jar:/Users/lee/.m2/repository/com/cognitect/transit-clj/1.0.324/transit-clj-1.0.324.jar:/Users/lee/.m2/repository/com/taoensso/timbre/5.1.2/timbre-5.1.2.jar:/Users/lee/.m2/repository/hiccup/hiccup/2.0.0-alpha2/hiccup-2.0.0-alpha2.jar:/Users/lee/.m2/repository/http-kit/http-kit/2.5.3/http-kit-2.5.3.jar:/Users/lee/.m2/repository/nrepl/bencode/1.1.0/bencode-1.1.0.jar:/Users/lee/.m2/repository/org/clojure/clojure/1.11.0-alpha1/clojure-1.11.0-alpha1.jar:/Users/lee/.m2/repository/org/clojure/core.async/1.3.618/core.async-1.3.618.jar:/Users/lee/.m2/repository/org/clojure/core.match/1.0.0/core.match-1.0.0.jar:/Users/lee/.m2/repository/org/clojure/data.csv/1.0.0/data.csv-1.0.0.jar:/Users/lee/.m2/repository/org/clojure/data.xml/0.2.0-alpha6/data.xml-0.2.0-alpha6.jar:/Users/lee/.m2/repository/org/clojure/test.check/1.1.0/test.check-1.1.0.jar:/Users/lee/.m2/repository/org/clojure/tools.cli/1.0.206/tools.cli-1.0.206.jar:/Users/lee/.m2/repository/org/clojure/tools.logging/1.1.0/tools.logging-1.1.0.jar:/Users/lee/.m2/repository/rewrite-clj/rewrite-clj/1.0.644-alpha/rewrite-clj-1.0.644-alpha.jar:/Users/lee/.m2/repository/selmer/selmer/1.12.44/selmer-1.12.44.jar:/Users/lee/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.12.4/jackson-core-2.12.4.jar:/Users/lee/.m2/repository/com/fasterxml/jackson/dataformat/jackson-dataformat-cbor/2.12.4/jackson-dataformat-cbor-2.12.4.jar:/Users/lee/.m2/repository/com/fasterxml/jackson/dataformat/jackson-dataformat-smile/2.12.4/jackson-dataformat-smile-2.12.4.jar:/Users/lee/.m2/repository/tigris/tigris/0.1.2/tigris-0.1.2.jar:/Users/lee/.m2/repository/org/flatland/ordered/1.5.9/ordered-1.5.9.jar:/Users/lee/.m2/repository/org/yaml/snakeyaml/1.26/snakeyaml-1.26.jar:/Users/lee/.m2/repository/com/cognitect/transit-java/1.0.343/transit-java-1.0.343.jar:/Users/lee/.m2/repository/com/taoensso/encore/3.12.1/encore-3.12.1.jar:/Users/lee/.m2/repository/io/aviso/pretty/0.1.37/pretty-0.1.37.jar:/Users/lee/.m2/repository/org/clojure/core.specs.alpha/0.2.56/core.specs.alpha-0.2.56.jar:/Users/lee/.m2/repository/org/clojure/spec.alpha/0.2.194/spec.alpha-0.2.194.jar:/Users/lee/.m2/repository/org/clojure/tools.analyzer.jvm/1.1.0/tools.analyzer.jvm-1.1.0.jar:/Users/lee/.m2/repository/org/clojure/data.codec/0.1.0/data.codec-0.1.0.jar:/Users/lee/.m2/repository/org/clojure/tools.reader/1.3.5/tools.reader-1.3.5.jar:/Users/lee/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.12.4/jackson-databind-2.12.4.jar:/Users/lee/.m2/repository/commons-codec/commons-codec/1.10/commons-codec-1.10.jar:/Users/lee/.m2/repository/javax/xml/bind/jaxb-api/2.3.0/jaxb-api-2.3.0.jar:/Users/lee/.m2/repository/org/msgpack/msgpack/0.6.12/msgpack-0.6.12.jar:/Users/lee/.m2/repository/com/taoensso/truss/1.6.0/truss-1.6.0.jar:/Users/lee/.m2/repository/org/clojure/core.memoize/1.0.236/core.memoize-1.0.236.jar:/Users/lee/.m2/repository/org/clojure/tools.analyzer/1.0.0/tools.analyzer-1.0.0.jar:/Users/lee/.m2/repository/org/ow2/asm/asm/5.2/asm-5.2.jar:/Users/lee/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.12.4/jackson-annotations-2.12.4.jar:/Users/lee/.m2/repository/com/googlecode/json-simple/json-simple/1.1.1/json-simple-1.1.1.jar:/Users/lee/.m2/repository/org/javassist/javassist/3.18.1-GA/javassist-3.18.1-GA.jar:/Users/lee/.m2/repository/org/clojure/core.cache/1.0.207/core.cache-1.0.207.jar:/Users/lee/.m2/repository/org/clojure/data.priority-map/1.0.0/data.priority-map-1.0.0.jar
```

### bb.edn :paths with aliases
But... if alias keywords are specified in `bb.edn` `:paths` they are not included in the classpath.
In a new empty directory, given the following `bb.edn`
```
{:paths ["src" "test" :aliasa :aliasb]
 :aliases {:aliasa ["a1" "a2"]
           :aliasb ["a3"]}}
```

Notice that babashka does not include `a1:a2:a3`:
```Shell
❯ bb print-deps --format classpath 
src:test:/Users/lee/.m2/repository/babashka/babashka.curl/0.0.3/babashka.curl-0.0.3.jar:/Users/lee/.m2/repository/babashka/fs/0.0.5/fs-0.0.5.jar:/Users/lee/.m2/repository/cheshire/cheshire/5.10.1/cheshire-5.10.1.jar:/Users/lee/.m2/repository/clj-commons/clj-yaml/0.7.107/clj-yaml-0.7.107.jar:/Users/lee/.m2/repository/com/cognitect/transit-clj/1.0.324/transit-clj-1.0.324.jar:/Users/lee/.m2/repository/com/taoensso/timbre/5.1.2/timbre-5.1.2.jar:/Users/lee/.m2/repository/hiccup/hiccup/2.0.0-alpha2/hiccup-2.0.0-alpha2.jar:/Users/lee/.m2/repository/http-kit/http-kit/2.5.3/http-kit-2.5.3.jar:/Users/lee/.m2/repository/nrepl/bencode/1.1.0/bencode-1.1.0.jar:/Users/lee/.m2/repository/org/clojure/clojure/1.11.0-alpha1/clojure-1.11.0-alpha1.jar:/Users/lee/.m2/repository/org/clojure/core.async/1.3.618/core.async-1.3.618.jar:/Users/lee/.m2/repository/org/clojure/core.match/1.0.0/core.match-1.0.0.jar:/Users/lee/.m2/repository/org/clojure/data.csv/1.0.0/data.csv-1.0.0.jar:/Users/lee/.m2/repository/org/clojure/data.xml/0.2.0-alpha6/data.xml-0.2.0-alpha6.jar:/Users/lee/.m2/repository/org/clojure/test.check/1.1.0/test.check-1.1.0.jar:/Users/lee/.m2/repository/org/clojure/tools.cli/1.0.206/tools.cli-1.0.206.jar:/Users/lee/.m2/repository/org/clojure/tools.logging/1.1.0/tools.logging-1.1.0.jar:/Users/lee/.m2/repository/rewrite-clj/rewrite-clj/1.0.644-alpha/rewrite-clj-1.0.644-alpha.jar:/Users/lee/.m2/repository/selmer/selmer/1.12.44/selmer-1.12.44.jar:/Users/lee/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.12.4/jackson-core-2.12.4.jar:/Users/lee/.m2/repository/com/fasterxml/jackson/dataformat/jackson-dataformat-cbor/2.12.4/jackson-dataformat-cbor-2.12.4.jar:/Users/lee/.m2/repository/com/fasterxml/jackson/dataformat/jackson-dataformat-smile/2.12.4/jackson-dataformat-smile-2.12.4.jar:/Users/lee/.m2/repository/tigris/tigris/0.1.2/tigris-0.1.2.jar:/Users/lee/.m2/repository/org/flatland/ordered/1.5.9/ordered-1.5.9.jar:/Users/lee/.m2/repository/org/yaml/snakeyaml/1.26/snakeyaml-1.26.jar:/Users/lee/.m2/repository/com/cognitect/transit-java/1.0.343/transit-java-1.0.343.jar:/Users/lee/.m2/repository/com/taoensso/encore/3.12.1/encore-3.12.1.jar:/Users/lee/.m2/repository/io/aviso/pretty/0.1.37/pretty-0.1.37.jar:/Users/lee/.m2/repository/org/clojure/core.specs.alpha/0.2.56/core.specs.alpha-0.2.56.jar:/Users/lee/.m2/repository/org/clojure/spec.alpha/0.2.194/spec.alpha-0.2.194.jar:/Users/lee/.m2/repository/org/clojure/tools.analyzer.jvm/1.1.0/tools.analyzer.jvm-1.1.0.jar:/Users/lee/.m2/repository/org/clojure/data.codec/0.1.0/data.codec-0.1.0.jar:/Users/lee/.m2/repository/org/clojure/tools.reader/1.3.5/tools.reader-1.3.5.jar:/Users/lee/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.12.4/jackson-databind-2.12.4.jar:/Users/lee/.m2/repository/commons-codec/commons-codec/1.10/commons-codec-1.10.jar:/Users/lee/.m2/repository/javax/xml/bind/jaxb-api/2.3.0/jaxb-api-2.3.0.jar:/Users/lee/.m2/repository/org/msgpack/msgpack/0.6.12/msgpack-0.6.12.jar:/Users/lee/.m2/repository/com/taoensso/truss/1.6.0/truss-1.6.0.jar:/Users/lee/.m2/repository/org/clojure/core.memoize/1.0.236/core.memoize-1.0.236.jar:/Users/lee/.m2/repository/org/clojure/tools.analyzer/1.0.0/tools.analyzer-1.0.0.jar:/Users/lee/.m2/repository/org/ow2/asm/asm/5.2/asm-5.2.jar:/Users/lee/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.12.4/jackson-annotations-2.12.4.jar:/Users/lee/.m2/repository/com/googlecode/json-simple/json-simple/1.1.1/json-simple-1.1.1.jar:/Users/lee/.m2/repository/org/javassist/javassist/3.18.1-GA/javassist-3.18.1-GA.jar:/Users/lee/.m2/repository/org/clojure/core.cache/1.0.207/core.cache-1.0.207.jar:/Users/lee/.m2/repository/org/clojure/data.priority-map/1.0.0/data.priority-map-1.0.0.jar
```

This probably makes sense - does Babashka even look at `bb.edn` `:aliases` at all right now?

### So...
I think that this means we should warn for any `:paths` vector elements that are not strings for `bb.edn` (this means we'll warn if `:paths` element is a keyword for `bb.edn`).  If/when we add `:paths` checking for `deps.edn` we'll allow for both keywords and strings.

Does that make sense for you too?

**Comment by borkdude** [comment]
I think that makes sense. My thought was to just re-use all the code we have for deps.edn, so if you are going to add paths checking for bb.edn then please just it once for both files.

**Comment by borkdude** [comment]
I mean some thing in in the deps-edn linting namespace can just be disabled based on the filename.

**Comment by lread** [comment]
Thanks, that helps.  I was being wary of feature-creeping past what I described in the issue.  
I will add appropriate `:paths` linting for both `bb.edn` and `deps.edn` as part of this lil' PR.

I shall follow up with a separate issue & PR to apply current `:deps.edn` `:deps` linting to `:bb.edn`.
